SpatialGDK/Source/SpatialGDK/Private/LoadBalancing/InterestManager.cpp

```diff
@@ -0,0 +1,650 @@
+				++RegionBox;
+			}
+			++Pos;
+			++Flags;
+			++VisibilityPtr;
+		}
+	}
+
+	const uint64 AllRegionsMask = [NumRegions] {
+		// To handle the 64 regions case, shifting by 64 could be a nop;
+		uint64 Mask = 1ull << (NumRegions - 1);
+		Mask <<= 1;
+		return Mask - 1ull;
+	}();
+
+	if (bVectorizeRegions)
+	{
+		SCOPE_CYCLE_COUNTER(STAT_InterestManagerComputationBoxSSE);
+
+		// SSE version, processing 4 regions at a time.
+		// 6ms, 200k entities, 64 regions
+		const uint32 Num4Regions = (NumRegions + 3) / 4;
+		const uint32 NextMultipleOf4 = Num4Regions * 4;
+
+		TArray<float, TAlignedHeapAllocator<16>> BoxesMinX;
+		TArray<float, TAlignedHeapAllocator<16>> BoxesMinY;
+		TArray<float, TAlignedHeapAllocator<16>> BoxesMaxX;
+		TArray<float, TAlignedHeapAllocator<16>> BoxesMaxY;
+		BoxesMinX.Reserve(NextMultipleOf4);
+		BoxesMinY.Reserve(NextMultipleOf4);
+		BoxesMaxX.Reserve(NextMultipleOf4);
+		BoxesMaxY.Reserve(NextMultipleOf4);
+		for (uint32 j = 0; j < NumRegions; ++j)
+		{
+			BoxesMinX.Add(Regions[j].Min.X);
+			BoxesMinY.Add(Regions[j].Min.Y);
+			BoxesMaxX.Add(Regions[j].Max.X);
+			BoxesMaxY.Add(Regions[j].Max.Y);
+		}
+		for (uint32 j = NumRegions; j < NextMultipleOf4; ++j)
+		{
+			BoxesMinX.Add(0);
+			BoxesMinY.Add(0);
+			BoxesMaxX.Add(0);
+			BoxesMaxY.Add(0);
+		}
+
+		const uint32* Flags = EntityFlags.GetData();
+		uint64* VisibilityPtr = Visibility.GetData();
+		const FVector2D* Pos = EntityPosition.GetData();
+		for (uint32 i = 0; i < NumEntities; ++i)
+		{
+			// Flags are currently used to control visibility of entity in all regions (eg. AlwaysRelevant), so if set, ensure visible to
+			// all regions.
+			*VisibilityPtr = ((uint64)(*Flags != 0)) * AllRegionsMask;
+
+			const float* BoxesMinXPtr = BoxesMinX.GetData();
+			const float* BoxesMinYPtr = BoxesMinY.GetData();
+			const float* BoxesMaxXPtr = BoxesMaxX.GetData();
+			const float* BoxesMaxYPtr = BoxesMaxY.GetData();
+
+			VectorRegister RegisterX = VectorSetFloat1(Pos->X);
+			VectorRegister RegisterY = VectorSetFloat1(Pos->Y);
+
+			for (uint32 j = 0; j < NextMultipleOf4 / 4; ++j)
+			{
+				VectorRegister BoxTest = VectorLoadAligned(BoxesMinXPtr);
+				VectorRegister TestResult = VectorCompareGT(RegisterX, BoxTest);
+
+				BoxTest = VectorLoadAligned(BoxesMaxXPtr);
+				TestResult = VectorBitwiseAnd(TestResult, VectorCompareGT(BoxTest, RegisterX));
+
+				BoxTest = VectorLoadAligned(BoxesMinYPtr);
+				TestResult = VectorBitwiseAnd(TestResult, VectorCompareGT(RegisterY, BoxTest));
+
+				BoxTest = VectorLoadAligned(BoxesMaxYPtr);
+				TestResult = VectorBitwiseAnd(TestResult, VectorCompareGT(BoxTest, RegisterY));
+
+				uint64 Mask = VectorMaskBits(TestResult);
+				*VisibilityPtr |= Mask << (j * 4);
+
+				BoxesMinXPtr += 4;
+				BoxesMinYPtr += 4;
+				BoxesMaxXPtr += 4;
+				BoxesMaxYPtr += 4;
+			}
+			++Pos;
+			++Flags;
+			++VisibilityPtr;
+		}
+	}
+// Disabled until we find how to make it work on linux.
```

Comment:
Any idea what is wrong with it on linux?

Reply:
AVX intrinsics

Reply:
worth wrapping in the comment?